src/blr/cli/tagbam.py

```diff
@@ -1,58 +1,100 @@
 """
-Transfers barcode sequence information from the input file alignment name to SAM tags in output file.
+Transfers SAM tags from query headers to SAM tags. Currently tags in header must follow SAM tag format, e.g.
+BC:Z:<SEQUENCE>.
 """
 
 import pysam
 import logging
+import re
 from tqdm import tqdm
+from collections import Counter
+
+from blr.utils import print_stats
 
 logger = logging.getLogger(__name__)
 
+ALLOWED_SAM_TAG_TYPES = "ABfHiZ"  # From SAM format specs https://samtools.github.io/hts-specs/SAMtags.pdf
+
 
 def main(args):
+    # Can't be at top since function are defined later
+    REGEX_FUNCTION_DICT = {
+        "sam": build_regex_sam_tag
+    }
 
-    # Generate dict with bc => bc_cluster consensus sequence
     logger.info("Starting analysis")
-    alignments_missing_bc = 0
+    summary = Counter()
```

Comment:
I usually define this outside the function scope e.g. like `logger`. Then it is accessible and can be updated in any function in the script.

Reply:
However, globals break encapsulation and you should avoid them whenever you can. For `logger`, it’s practical (and it’s designed to be used that way), but for statistics, it isn’t too hard to keep them local. Try to design the script you’re writing as if you were designing a library. For libraries, global state is evil because it makes them harder to use. You’d get incorrect statistics if you called the `main()` function twice.

Reply:
I guess it is not too difficult to keep it local. Then we should probably change this in the other script where this is implemented.